base/set.jl

```diff
@@ -8,6 +8,11 @@ type Set{T} <: AbstractSet{T}
 end
 Set() = Set{Any}()
 Set(itr) = Set{eltype(itr)}(itr)
+function Set(g::Generator)
+    T = _default_eltype(typeof(g))
+    T === Union{} && return Set(collect(g))
```

Comment:
I believe we can use `grow_to!(Set{Union{}}(), itr)` to avoid calling `collect`.